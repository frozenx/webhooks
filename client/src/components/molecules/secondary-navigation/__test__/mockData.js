import React from 'react';
import { Link } from 'react-router-dom';

const expectedMenuItems = [
    {
      id: "managesupplier",
      hasPopup: false,
      altLink: () => (
        <Link to="/tradingpartner/manage">Manage suppliers</Link>
      ),
      text: "Manage suppliers"
    },
    {
      id: "createsupplier",
      hasPopup: false,
      altLink: () => <Link to="/tradingpartner">Create new supplier</Link>,
      text: "Create new supplier"
    }
  ]

  export default {
      expectedMenuItems
  }