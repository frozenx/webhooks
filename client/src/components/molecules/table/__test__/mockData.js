const mockSiteColumns = [null, 'column head 1'];
const mockSiteRows = [{ siteName: 'some site name', siteCodeId: '123' }];


const mockPaymentColumns = [null, 'column head 1'];
const mockPaymentRows = [{ paymentTermsName: '123' }];


export default {
  mockSiteColumns,
  mockSiteRows,
  mockPaymentColumns,
  mockPaymentRows,
};
