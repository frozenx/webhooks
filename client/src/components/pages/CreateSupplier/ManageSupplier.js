import React from 'react';
import Search from '../search';

const manageSupplier = (props) => {
    return (
        <Search {...props}/>
    )
}

export default manageSupplier;