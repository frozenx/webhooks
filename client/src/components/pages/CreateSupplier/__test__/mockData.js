export const foundSupplier = [
  {
  "entityType": "partner",
  "uuid": "d682662a-7642-31d8-a7c1-e1dd1ddc7c47",
  "number": "1752657121",
  "name": "TEST98",
  "taxOrganisationType": "Corporation"
}]

export const mockFormData = {
  "partnerName": "TEST98",
  "companyRegNumber": "122",
  "vatRegNumber": "22"
}

export const mockResponse = {
    "attributeGroups": [
      {
        "id": "trn:tesco:partner:attributegroup:0067b034-b795-451c-b738-16f40330c339",
        "name": "Partner Details",
        "behaviour": "Single Row",
        "type": "Dynamic",
        "key": "partner",
        "attributes": [         
          {
            "enabled": true,
            "required": true,
            "type": "String",
            "manifest": "Editable",
            "value": "",
            "key": "partnerName",
            "id": "trn:tesco:partner:attribute:042d8298-bb82-4085-bf55-202964ab6145",
            "displayName": "Partner legally registered name",
            "name": "Partner legally registered name",
            "toolTip": "Tooltip not needed here",
            "rules": [],
            "options": [
              {
                "value": "True",
                "displayValue": "Yes",
                "lang": "en"
              },
              {
                "value": "False",
                "displayValue": "No",
                "lang": "en"
              }
            ],
            "regexRule": "",
            "attributeGroupKey": "partner"
          }
        ]
      },     
    ],
    "attributeToRulesMapping": {
      "uuid": {
        "regexRule": "",
        "required": false,
        "attrGrpKey": "partner",
        "read": false,
        "create": false,
        "update": false
      },
      "partnerNumber": {
        "regexRule": "",
        "required": false,
        "attrGrpKey": "partner",
        "read": true
      },
      "partnerName": {
        "regexRule": "",
        "required": true,
        "attrGrpKey": "partner",
        "read": true,
        "create": true,
        "update": true
      },      
      "companyRegNumber": {
        "regexRule": "",
        "required": false,
        "attrGrpKey": "partner",
        "read": true,
        "create": true,
        "update": true
      },

      "vatRegNumber": {
        "regexRule": "",
        "required": true,
        "attrGrpKey": "partner",
        "read": true,
        "create": true,
        "update": true
      },
           
    },
    "partnerAttributeStatic": {
      "attributes": [
        {
          "enabled": true,
          "required": true,
          "type": "Number",
          "manifest": "Read Only",
          "value": "",
          "key": "partnerNumber",
          "id": "trn:tesco:partner:attribute:f7284ff8-b4b8-451d-a0fb-5153e2b97595",
          "displayName": "Partner number",
          "name": "Partner number",
          "toolTip": "Tooltip not needed here",
          "rules": [
            {
              "uuid": "trn:tesco:partner:attributerule:19e75447-a0a1-4293-a8cb-6165084fc4c1",
              "ruleType": "regex",
              "definition": "^\\d{10}$",
              "messages": [
                {
                  "message": "Invalid Partner Number. Length should be 10 digit length",
                  "lang": "en"
                }
              ],
              "message": "Invalid Partner Number. Length should be 10 digit length"
            }
          ],
          "options": [
            {
              "value": "True",
              "displayValue": "Yes",
              "lang": "en"
            },
            {
              "value": "False",
              "displayValue": "No",
              "lang": "en"
            }
          ],
          "regexRule": "^\\d{10}$",
          "attributeGroupKey": "partner"
        },
        {
          "enabled": true,
          "required": true,
          "type": "String",
          "manifest": "Editable",
          "value": "",
          "key": "partnerName",
          "id": "trn:tesco:partner:attribute:042d8298-bb82-4085-bf55-202964ab6145",
          "displayName": "Partner legally registered name",
          "name": "Partner legally registered name",
          "toolTip": "Tooltip not needed here",
          "rules": [],
          "options": [
            {
              "value": "True",
              "displayValue": "Yes",
              "lang": "en"
            },
            {
              "value": "False",
              "displayValue": "No",
              "lang": "en"
            }
          ],
          "regexRule": "",
          "attributeGroupKey": "partner"
        },
        {
          "enabled": true,
          "required": false,
          "type": "Number",
          "manifest": "Editable",
          "value": "",
          "key": "companyRegNumber",
          "id": "trn:tesco:partner:attribute:f175fc1a-06b2-402c-8d2a-828fdba77aeb",
          "displayName": "Company registration number",
          "name": "Company registration number",
          "toolTip": "Tooltip not needed here",
          "rules": [],
          "options": [
            {
              "value": "True",
              "displayValue": "Yes",
              "lang": "en"
            },
            {
              "value": "False",
              "displayValue": "No",
              "lang": "en"
            }
          ],
          "regexRule": "",
          "attributeGroupKey": "partner"
        },
        {
          "enabled": true,
          "required": true,
          "type": "String",
          "manifest": "Editable",
          "value": "",
          "key": "vatRegNumber",
          "id": "trn:tesco:partner:attribute:61b1e1e5-4678-4a2b-9a8b-914c7b4e0979",
          "displayName": "VAT number",
          "name": "VAT number",
          "toolTip": "Tooltip not needed here",
          "rules": [],
          "options": [
            {
              "value": "True",
              "displayValue": "Yes",
              "lang": "en"
            },
            {
              "value": "False",
              "displayValue": "No",
              "lang": "en"
            }
          ],
          "regexRule": "",
          "attributeGroupKey": "partner"
        }
      ]
    }
  }