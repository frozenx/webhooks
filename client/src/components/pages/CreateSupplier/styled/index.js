import styled from 'styled-components'

export const CreateSupplierWrapper = styled.div`
    .notification{
        margin-bottom: 34px;
    }
`