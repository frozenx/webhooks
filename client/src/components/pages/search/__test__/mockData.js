const mockUamContext = {
  resources: {
    api: [{ '/supplier': ['view'] }, { '/supplier-site': ['view'] }],
    page: {
      '/': {
        tabs: [{ managesupplier: ['view'] }, { createsupplier: ['view'] }],
      },
      supplierlist: { button: [{ select: ['enable'] }] },
    },
  },
};

const mockUamContextWithEmptyAccessForRootPage = {
  resources: {
    page: {
      '/': null,
    },
  },
};

const mockUamContextWithFailureDataForRootPage = {
  resources: {
    page: {
      '/': {
        tabs: {},
      },
    },
  },
};

export default {
  mockUamContext,
  mockUamContextWithEmptyAccessForRootPage,
  mockUamContextWithFailureDataForRootPage,
};
