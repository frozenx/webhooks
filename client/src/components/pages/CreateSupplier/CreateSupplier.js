import React, { Component } from "react";
import { FormHeading, SecondaryWrapper, NotificationStyle } from "../../../common/styles";
import { CreateSupplierWrapper } from "./styled";
import { LanguagePack } from '../../context';
import Button from "@beans/button";
import Notification from "@beans/notification";
import GenericFormContainer from "../../molecules/GenericForm/container";
import config from '../../../config/';
import httpClient from '../../../lib/httpClient';
import Loader from "../../../components/atoms/Loader";
import constants from './constants'

class CreateSupplier extends Component {
  state = {
    values: {},
    errorStatus: {},
    regex: {},
    required: {},
    submitButtonStatus: false,
    supplierHeaderData: [],
    showStepTwo: false,
    showStepThree: false,
    showStepOne: true,
    displayLoader: false,
    displayErrorStatus: false,
    displayErrorMessage: null,
    supplierMetaDataHeader: null,
    partnerName: null
  };

  async componentWillMount() {
    const { partnerNumber } = constants
    this.setState({
      displayLoader: true
    });
    try {
      const response = await httpClient.get(
        config.endPoints.supplierHeaderEndPoint
      );
      if (response.status > 400) {
        const customError = {
          message: response.message
        };
        throw customError;
      }
      const filteredSupplierMetaDataHeader = response.partnerAttributeStatic.attributes.filter((attribute) => attribute.key !== partnerNumber);
      this.setState({
        displayLoader: false,
        supplierMetaDataHeader: filteredSupplierMetaDataHeader,
        attributeRules: response.attributeToRulesMapping
      });
    } catch (error) {
      this.setState({
        displayLoader: false,
        displayErrorStatus: true,
        displayErrorMessage: error.message
      });
    }

    let selectedAttr, selectedAttrErrStatus, selectedAttrRegex, selectedAttrRequired;
    const { supplierMetaDataHeader, attributeRules } = this.state;
    if (supplierMetaDataHeader && supplierMetaDataHeader.length > 0) {
      supplierMetaDataHeader.forEach(attr => {
        selectedAttr = { ...selectedAttr };
        selectedAttr[attr.key] = attr.value;
        selectedAttrErrStatus = { ...selectedAttrErrStatus };
        selectedAttrErrStatus[attr.key] = false;
        selectedAttrRegex = { ...selectedAttrRegex };
        selectedAttrRegex[attr.key] = attributeRules[attr.key].regexRule;
        selectedAttrRequired = { ...selectedAttrRequired };
        selectedAttrRequired[attr.key] = attributeRules[attr.key].required;
        return { selectedAttr, selectedAttrErrStatus, selectedAttrRegex, selectedAttrRequired };
      });
      this.setState({
        values: selectedAttr,
        errorStatus: selectedAttrErrStatus,
        regex: selectedAttrRegex,
        required: selectedAttrRequired,
        createSupplierData: supplierMetaDataHeader
      });
    }
  }


  _handleChange = e => {
    const initalValues = {};
    const { id, value } = e.target;
    if (config.forceCapitalizationAttributes[id]) {
      initalValues[id] = value.toUpperCase();
    } else {
      initalValues[id] = value
    }
    const initialErrorStatus = {};
    const initialRegex = { ...this.state.regex };
    const initialRequired = { ...this.state.required };
    if ((!value && initialRequired[id]) || !new RegExp(initialRegex[id]).test(value)) {
      initialErrorStatus[id] = true;
    } else {
      initialErrorStatus[id] = false
    }

    this.setState(prevState => ({
      isSupplierFound: false,
      values: { ...prevState.values, ...initalValues },
      errorStatus: { ...prevState.errorStatus, ...initialErrorStatus }
    }))

  }

  _handleValidation() {
    let tempValues = { ...this.state.values };
    let tempErrStatus = { ...this.state.errorStatus };
    let tempRegex = { ...this.state.regex };
    let tempRequired = { ...this.state.required };
    let formIsValid = true;
    Object.keys(tempValues).forEach(key => {
      if ((!tempValues[key] && tempRequired[key]) || !new RegExp(tempRegex[key]).test(tempValues[key])) {
        tempErrStatus[key] = true;
        formIsValid = false;
      } else {
        tempErrStatus[key] = false;
      }
    });

    this.setState(prevState => ({
      errorStatus: { ...prevState.errorStatus, ...tempErrStatus }
    }));

    return formIsValid;
  }

  _handleSubmit = async e => {
    e.preventDefault();
    let validate = this._handleValidation();
    if (validate) {
      const formData = this.state.values;      
      const searchSupplierUrl = `${config.endPoints.searchAndCreate}`
      this.setState({ displayLoader: true });
      try {
        const searchResults = await httpClient.post(searchSupplierUrl, formData);
        if (searchResults.suppliers) {
          this.setState({ isSupplierExists: true, isCompanyAddressError: false, searchResults: searchResults })
          setTimeout(() => {
            this.props.history.push({ pathname: '/tradingpartner', state: { supplierResults: this.state.searchResults, searchedPartnerName: formData.partnerName } })
          }, 1000)
        } else if (searchResults.errorKeys) {
          const errors = searchResults.errorKeys.reduce((errorObject, errorKey) => {
            return { ...errorObject, [errorKey]: true }
          }, { ...this.state.errorStatus });
          this.setState({
            errorStatus: { ...errors },
            isCompanyAddressError: true,
            displayLoader: false,
          });
        } else {
          this.props.history.push({ pathname: '/tradingpartner/supplier', state: { supplierHeaderData: searchResults } })
          this.setState({
            displayLoader: false,
            showStepOne: false,
            showStepTwo: true,
            supplierHeaderData: searchResults,
          });
        }
      } catch (err) {
      }
    } else {
    }
  };

  _handleAttributeChange = () => {}

  render() {
    const {
      displayLoader,
      values,
      errorStatus,
      showStepOne,
      isCompanyAddressError,
      isSupplierExists,
      createSupplierData,
    } = this.state;

    const i18Data = this.context;
    return (
      <>
        {displayLoader ? <Loader /> : null}
        {showStepOne ? (
          <SecondaryWrapper>
            {
              createSupplierData && createSupplierData.length > 0 ?
                <CreateSupplierWrapper>
                  <FormHeading>
                    <h1>{i18Data.CREATE_NEW_SUPPLIER_TEXT}</h1>
                    <h4>{i18Data.CREATE_NEW_SUPPLIER_SUBTEXT}</h4>
                  </FormHeading>
                  <form
                    onChange={e => this._handleChange(e)}
                  >
                    {isSupplierExists &&
                      <NotificationStyle>
                        <Notification title={'Supplier already exists'} variant={'error'} className='notification' />
                      </NotificationStyle>
                    }
                    {
                      isCompanyAddressError &&
                      <NotificationStyle>
                        <Notification title={'Company number entered is invalid'} variant={'error'} className='notification' />
                      </NotificationStyle>
                    }
                    <GenericFormContainer
                      genericFormData={this.state.createSupplierData}
                      values={values}
                      errors={errorStatus}
                      submitButtonStatus={this.state.submitButtonStatus}
                      attributeToRulesMapping={this.state.attributeRules}
                      handleAttributeChange={this._handleAttributeChange}
                    />
                    <Button
                      variant="primary"
                      className={`submit-btn`}
                      type="submit"
                      onClick={this._handleSubmit}>
                      {i18Data.CONFIRM_BTN}
                    </Button>
                  </form>
                </CreateSupplierWrapper>
                :
                null
            }
          </SecondaryWrapper>
        )
          : null}
      </>
    );
  }
}

CreateSupplier.contextType = LanguagePack;

export default CreateSupplier;
