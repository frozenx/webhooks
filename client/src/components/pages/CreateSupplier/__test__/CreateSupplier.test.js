jest.mock('../../../../lib/httpClient')
import React from 'react';
import {shallow} from 'enzyme';
import CreateSupplier from '../CreateSupplier';
import renderer from 'react-test-renderer';
import httpClient from '../../../../lib/httpClient';
import { DefaultThemeProvider } from '@beans/theme';
import { mockResponse, mockFormData, foundSupplier } from './mockData'


describe("Creat Supplier test cases", () => {
  let Wrapper,
      tree,
      instance;
  let mockHistory= {
      push: jest.fn()
  } 

  const response = mockResponse

  beforeEach(() => {
    tree = renderer.create(
      <DefaultThemeProvider>
        <CreateSupplier history={mockHistory}/>
      </DefaultThemeProvider>
    ).toJSON();

    httpClient.get.mockResolvedValue(response);
    Wrapper = shallow(<CreateSupplier history={mockHistory}/>)
    instance = Wrapper.instance();

  });

  test("renders the create supplier component", () => {
    expect(tree).toMatchSnapshot();
  });
  test("should show the error status when api fails in", (done) => {
    let mockResponse = {
      status: 500,
      message: 'Error'
    }
    httpClient.get.mockResolvedValue(mockResponse);
    Wrapper = shallow(<CreateSupplier history={mockHistory}/>)
    instance = Wrapper.instance();
    setTimeout(() => {
      expect(Wrapper.state().displayErrorStatus).toBe(true);
      expect(Wrapper.state().displayErrorMessage).toBe(mockResponse.message)
      done();
    }, 0);
  });
  describe("form actions", () => {    
    test("should expose a method for onChange", (done) => {
        jest.spyOn(instance, '_handleChange');

      setTimeout(() => {
        Wrapper.find('form').simulate("change", { target: {  id:'partnerName', value: "123" } });
        expect(instance._handleChange).toHaveBeenCalled()        
        expect(Wrapper.state().isSupplierFound).toBe(false);      
        done();
      }, 0);
    });
    test("should render the error status onChange when value is empty", (done) => {
      Wrapper.setState({required : {name : true}})
      setTimeout(() => {
        Wrapper.find('form').simulate("change", { target: { id: "name", value: "" } });
        expect(Wrapper.state().errorStatus['name']).toBe(true)
        done();
      }, 0);
    });
  });
  describe("onSubmit of form", () => {
    let event;
    beforeEach(() => {      
      event = { preventDefault: jest.fn() };
    });
    test("should call _handleSubmit method and _handleValidation method and make a success on validating the fields", () => {
      jest.spyOn(instance, '_handleSubmit');
      jest.spyOn(instance, '_handleValidation');
      Wrapper.setState({values: mockFormData})
      Wrapper.find('.submit-btn').simulate('click', event);
      expect(instance._handleSubmit).toHaveBeenCalledWith(event)
      expect(instance._handleValidation).toHaveBeenCalled()
    });

    test("should show the error status when value is empty", () => {
      Wrapper.setState({required : {name : true}, values: mockFormData})
      Wrapper.find('form').simulate("change", { target: { id: "name", value: "" } });
      Wrapper.find('.submit-btn').simulate('click', event);
      expect(Wrapper.state().errorStatus['name']).toBe(true)
    });
    
    
    test("should throw a server error", (done) => {
      Wrapper.setState({values: mockFormData})
      let mockErrorResponse = {
        suppliers: null
      };
      httpClient.post.mockResolvedValue(mockErrorResponse)
      Wrapper.find('.submit-btn').simulate('click', event);
      setTimeout(() => {
        expect(Wrapper.state().showStepTwo).toBe(true)
        expect(Wrapper.state().showStepOne).toBe(false)
        expect(Wrapper.state().supplierHeaderData.suppliers).toBe(null)
        done()
      },0)
    });
    test("should show the error when the search results has an error", (done) => {
      Wrapper.setState({values: mockFormData})
      let mockErrorResponse = {
          errorKeys: ['companyRegNumber']        
      };
      httpClient.post.mockResolvedValue(mockErrorResponse)
      Wrapper.find('.submit-btn').simulate('click', event);
      setTimeout(() => {
        expect(Wrapper.state().errorStatus['companyRegNumber']).toBe(true)
        done()
      },0)
    });
    test("should show the results if the supplier is alredy created", async(done) => {
      Wrapper.setState({values: mockFormData})
      let mockErrorResponse = {
        suppliers: foundSupplier
      };
      httpClient.post.mockResolvedValue(mockErrorResponse)
      Wrapper.find('.submit-btn').simulate('click', event);
      setTimeout(() => {
        expect(Wrapper.state().isSupplierExists).toBe(true)
        expect(Wrapper.state().searchResults.suppliers).toEqual(foundSupplier)
        done()
      },2000)
    });
  });
});
