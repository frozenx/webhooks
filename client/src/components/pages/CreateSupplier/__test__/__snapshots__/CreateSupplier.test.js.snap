// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Creat Supplier test cases renders the create supplier component 1`] = `
Array [
  <div
    className="sc-bdVaJa kybHWQ"
  >
    <span
      className="sc-bwzfXH eDyLnt"
    />
  </div>,
  <div
    className="base-components__RootElement-sc-150tvch-1 sc-bxivhb kUxGyj"
  />,
]
`;
